types/retry-handler.d.ts

```diff
@@ -0,0 +1,109 @@
+import Dispatcher from "./dispatcher";
+
+export default RetryHandler;
+
+declare class RetryHandler implements Dispatcher.DispatchHandlers {
+  constructor(
+    options: Dispatcher.DispatchOptions,
+    retryHandlers: RetryHandler.RetryHandlers,
+    retryOptions?: RetryHandler.RetryOptions
+  );
+}
+
+declare namespace RetryHandler {
+  export type RetryState = { counter: number; currentTimeout: number };
+
+  export type RetryCallback = (
+    err: Error,
+    state: RetryState,
+    opts: RetryOptions
+  ) => number | null;
+
+  export interface RetryOptions {
+    /**
+     * Callback to be invoked on every retry iteration.
+     * It receives the error, current state of the retry object and the options object
+     * passed when instantiating the retry handler.
+     *
+     * @type {RetryCallback}
+     * @memberof RetryOptions
+     */
+    retry?: RetryCallback;
```

Comment:
All the options below are only relevant if `retry` is null. Maybe this is better represented as an union?

Reply:
Kind of, they `retryOpts` are also passed down to the custom `retry` fn; the goal was to allow them also make use of them if they want, so the settings they set, can also be consumed by the same implementation; I can see the potential but do not have strong opinion wether to keep it or not, thoughts?